auth_mellon_util.c

```diff
@@ -1540,7 +1612,11 @@ bool am_validate_paos_header(request_rec *r, const char *header)
         goto cleanup;
     }
 
-    /* Split the header on the semicolon character */
+    /*
+     * Split the header on the semicolon character.
+     * There should be 2 items, the first is the version delclaration
+     * and the second is the service and it's options.
+     */
     semicolon_tokens = g_strsplit(header, ";", 0);
```

Comment:
Back when I first started working on this I checked Apache to see if there was utility functions I could call to parse the header, there wasn't. But I checked again to be sure. Sadly there is no common code, it appears as if every module implements their own hand crafted code. I suppose that shouldn't be surprising because HTTP header fields are quirky and non-uniform, every header seems to be just a little bit different.

But anyway I rewrote the parser to tokenize and then analyze the token sequence. It has very good diagnostics and will report the exact location in the string where it detected a problem. I tested the parser in stand-alone mode it handled anything I threw at it.